clients/src/main/java/org/apache/kafka/common/protocol/Protocol.java

```diff
@@ -1199,6 +1199,168 @@
 
     public static final Schema[] INIT_PRODUCER_ID_RESPONSE = new Schema[] {INIT_PRODUCER_ID_RESPONSE_V0};
 
+    public static final Schema ADD_PARTITIONS_TO_TXN_REQUEST_V0 = new Schema(
+            new Field("transactional_id",
+                    STRING,
+                    "The transactional id corresponding to the transaction."),
+            new Field("pid",
+                    INT64,
+                    "Current PID in use by the transactional id"),
+            new Field("epoch",
+                    INT16,
+                    "Current epoch associated with the PID"),
+            new Field("topic_partitions",
```

Comment:
It depends on how you look at it. `topic_partitions` implies a flat array of topic partitions. But, for efficiency, we generally implement an array of topics where each topic has the associated partitions and any other data associated with each partition. To illustrate what I mean in Scala:

```scala
val topics: Seq[Topic]

case class Topic(name: String, partitions: Seq[Partition])
```

I think that's the reasoning behind the naming schema that has been used so far. It's not perfect, but seems justifiable. So, I'd personally stick with it. Other in-progress PRs like KIP-101 do that.